server/filestore.go

```diff
@@ -3095,6 +3098,21 @@ func (mb *msgBlock) writeMsgRecord(rl, seq uint64, subj string, mhdr, msg []byte
 		}
 	}
 
+	// Check if we are tracking per subject for our simple state.
+	// Do this before changing the cache that would trigger a flush pending msgs call
+	// if we needed to regenerate the per subject info.
+	if len(subj) > 0 && !mb.noTrack {
+		if err := mb.ensurePerSubjectInfoLoaded(); err != nil {
+			return err
+		}
+		if ss := mb.fss[subj]; ss != nil {
+			ss.Msgs++
+			ss.Last = seq
```

Comment:
Do we need to check that `seq >= ss.Last` before doing the assignment or are we sure that it always be that writeMsgRecord is called with the "last" sequence?

Reply:
It is always called with the last sequence, or next to be more exact, and is locked.